### PR TITLE
Add functional tests

### LINKED ISSUE #78: Use examples as functional tests
Ooh right, so according to the plan above,
* Unit tests will be moved from `tests/` to `tests/unit/`
* Audio will be in `tests/`.  In this same dir, we create a `conftest.py` with a fixture of `basename(__file__)/audiofile` (use `os.path.{join,basename}` for this)
* Buffer unit test, i.e. in `tests/unit/test_buffer.py`, can access to the audio paths as the form of fixture, e.g. `def test_cache(context, mp3): cache([mp3])`
* Functional tests can also access the same audios, but run the examples instead, i.e. use `subprocess` to call, e.g. `[sys.executable, 'palace-hrtf.py', flac]`

_Originally posted by @McSinyx in https://github.com/McSinyx/palace/issues/39#issuecomment-616299660_

### LINKED ISSUE #80: Tox can't import numpy
When I run tox, the test for tonegen failed, because it can't import numpy.

> ModuleNotFoundError: No module named 'numpy'

I'm not sure if this is only a problem with tox, or we can change some parameters in tox.ini, say, so that it can import numpy.

If this is not fixed, probably we can only ignore tonegen for now, or to write tonegen without numpy. I'm not sure if it can import scipy, though.

### PR DESCRIPTION
**Pull request by Huy-Ngo**
This PR resolves #78 : Move current tests to `unit/` and add functional tests, which use example files, to `functional/`.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -1,38 +1,5 @@
...
-# by the Free Software Foundation, either version 3 of the License,
-# or (at your option) any later version.
-#
-# palace is distributed in the hope that it will be useful,
-# but WITHOUT ANY WARRANTY; without even the implied warranty of
-# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-# GNU Lesser General Public License for more details.
-#
-# You should have received a copy of the GNU Lesser General Public License
-# along with palace.  If not, see <https://www.gnu.org/licenses/>.
```
Comment (McSinyx) [inline tests/conftest.py:17]: Please give this one a header (with your name of course, I haven't touched this).

File: tests/functional/test_examples.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from subprocess import run
```
Comment (McSinyx) [inline tests/functional/test_examples.py:22]: Also LGPL header please.

**Review by McSinyx — CHANGES_REQUESTED** [review]
Some really early review.  Also dumb dumb git(hub? not sure sume to blame), shouldof shown the conftest as moved

File: tests/conftest.py [hunk 3]
```diff
@@ -1,5 +1,5 @@
-# test environment
-# Copyright (C) 2020  Nguyễn Gia Phong
+# Source pytest module
```
Comment (McSinyx) [inline tests/conftest.py]: Typo here

File: tests/functional/test_examples.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
+# Source pytest module
```
Comment (McSinyx) [inline tests/functional/test_examples.py]: Here too

**Review by McSinyx — CHANGES_REQUESTED** [review]
Just a simple reminder that it'd be nice if test funxons have docstrings.  BTW `tests/unit/test_context.py` also has mismatched description I thing.

File: tests/functional/test_examples.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ def test_hrtf():
 
 
 def test_info():
-    info = run([executable, './palace-hrtf.py', WAV], capture_output=True)
+    info = run([executable, './palace-hrtf.py'], capture_output=True)
```
Comment (McSinyx) [inline tests/functional/test_examples.py]: Oh no, why?

**Comment by McSinyx** [comment]
BTW could we have path to examples as constants too, please?

File: tests/functional/test_examples.py (hunk 5 above)
Comment (Huy-Ngo) [reply tests/functional/test_examples.py]: It's supposed to be running `palace-info.py`. I was copying so there was a lot of mistakes. In the current one I think it's fine.

**Comment by Huy-Ngo** [comment]
Hmm... flake8 doesn't recognize WAV defined in conftest.py

Is that fine? How to fix that?

File: tests/functional/test_examples.py [hunk 6]
```diff
@@ -19,21 +19,30 @@
 from subprocess import run
 from sys import excecutable
 
+EVENT = './palace-event.py'
+HRTF = './palace-hrtf.py'
+INFO = './palace-info.py'
+LATENCY = './palace-latency.py'
+REVERB = '/palace-reverb.py'
+STDEC = './palace-stdec.py'
+TONEGEN = '/palace-tonegen.py'
```
Comment (McSinyx) [inline tests/functional/test_examples.py]: As `join(basename(...` please, or even better: `join(abspath(basename(...` (same for the audios).

File: tests/functional/test_examples.py [hunk 7]
```diff
@@ -46,33 +55,32 @@ def test_info():
...
     assert 'Opened' in stdec.stdout
     assert 'Playing' in stdec.stdout
 
 
 def test_tonegen():
-    waveforms = ['sine', 'square', 'sawtooth', 'triangle', 'impulse', 'white-noise']
-    for waveform in waveforms:
-        tonegen = run([executable, '/palace-tonegen.py', '-w', waveform], capture_output=True)
+    for waveform in WAVEFORMS:
+        tonegen = run([executable, TONEGEN, '-w', waveform], capture_output=True)
```
Comment (McSinyx) [inline tests/functional/test_examples.py]: the pytest way is to [parametrize](https://docs.pytest.org/en/latest/parametrize.html)

**Comment by McSinyx** [comment]
> Hmm... flake8 doesn't recognize WAV defined in conftest.py

For now add a `# noqa` at the end of the reported line.  We'll figure out sth out later (not today).

File: tests/functional/test_examples.py (hunk 6 above)
Comment (Huy-Ngo) [reply tests/functional/test_examples.py]: Why do we have to do this? What are we joining the path with?

As for the audios, you mean the ones defined in `tests/conftest.py`? Isn't it properly path-ed yet?

File: tests/functional/test_examples.py [hunk 8]
```diff
@@ -16,28 +16,29 @@
...
-REVERB = '/palace-reverb.py'
-STDEC = './palace-stdec.py'
-TONEGEN = '/palace-tonegen.py'
+EVENT = join(abspath(basename('./palace-event.py')))
+HRTF = join(abspath(basename('./palace-hrtf.py')))
+INFO = join(abspath(basename('./palace-info.py')))
+LATENCY = join(abspath(basename('./palace-latency.py')))
+REVERB = join(abspath(basename('./palace-reverb.py')))
+STDEC = join(abspath(basename('./palace-stdec.py')))
+TONEGEN = join(abspath(basename('./palace-tonegen.py')))
```
Comment (McSinyx) [inline tests/functional/test_examples.py]: HAI

Should have said it clearer, I meant `join(abspath(basename(__file__)), ...)`.

Please do the same for the audio.

KTHXBYE

File: tests/functional/test_examples.py (hunk 6 above)
Comment (McSinyx) [reply tests/functional/test_examples.py]: See the other review below.  IMHO absolute path is a bit safer, could be wrong tho.

File: tests/functional/test_examples.py (hunk 8 above)
Comment (Huy-Ngo) [reply tests/functional/test_examples.py]: Wouldn't `basename(__file__)` return `test_examples.py`? I'm confused.

Comment (Huy-Ngo) [reply tests/functional/test_examples.py]: OK I will use `dirname(__file__)` instead since that's the only way that would make sense to me.

Comment (McSinyx) [reply tests/functional/test_examples.py]: LOL yes, my mistake!  Sorry for the confusions.

**Comment by Huy-Ngo** [comment]
Oh fkn no some checks were not successful >.>

**Review by McSinyx — APPROVED** [review]

**Pull request merged**